examples/pytorch/rgcn/link_predict_mp.py

```diff
@@ -0,0 +1,1147 @@
+
+        # only half of the edges will be used as graph structure
+        pos_idx = np.random.choice(bsize, size=bsize//2, replace=False)
+        pos_idx = th.tensor(pos_idx).long()
+        subg = dgl.remove_edges(subg, pos_idx)
+        rels = subg.edata['etype']
+
+        subg = dgl.add_reverse_edges(subg, copy_ndata=True)
+        # dgl.NID and dgl.EID are not copy automatically
+        subg.ndata['old_nid'] = org_nid
+        subg.edata['etype'] = th.cat([rels, rels + self.num_rel])
+
+        # calculate norm for compact_subg
+        in_deg = subg.in_degrees(range(subg.number_of_nodes())).float()
+        norm = 1.0 / in_deg
+        norm[th.isinf(norm)] = 0.
+        subg.ndata['norm'] = norm
+        subg.apply_edges(lambda edges : {'norm' : edges.dst['norm']})
+        subg.edata['norm'] = subg.edata['norm'].unsqueeze(1)
+        return (bsize, subg, p_u, p_v, p_rel)
+
+class LinkNeighborSampler:
+    """ Neighbor based sampler.
+
+    In Neighbor sampler, we randomly select K positive edges and N negative edges.
+    n-hop neighbor sample is applied to both positive nodes and negative nodes to
+    generate p_blocks and n_blocks.
+    """
+    def __init__(self, g, neg_edges, num_edges, num_neg, fanouts,
+        is_train=True, keep_pos_edges=False):
+        self.g = g
+        self.neg_edges = neg_edges
+        self.num_edges = num_edges
+        self.num_neg = num_neg
+        self.fanouts = fanouts
+        self.is_train = is_train
+        self.keep_pos_edges = keep_pos_edges
+
+    def sample_blocks(self, seeds):
+        pseeds = th.tensor(seeds).long()
+        bsize = pseeds.shape[0]
+        if self.num_neg is not None:
+            nseeds = th.randint(self.num_edges, (self.num_neg,))
+            nseeds = self.neg_edges[nseeds]
+        else:
+            nseeds = th.randint(self.num_edges, (bsize,))
+            nseeds = self.neg_edges[nseeds]
+
+        g = self.g
+        fanouts = self.fanouts
+        assert len(g.canonical_etypes) == 1
+        p_g = g.edge_subgraph(pseeds)
+        n_g = g.edge_subgraph(nseeds)
+
+        pg_seed = p_g.ndata[dgl.NID]
+        ng_seed = n_g.ndata[dgl.NID]
+
+        p_blocks = []
+        n_blocks = []
+        p_curr = pg_seed
+        n_curr = ng_seed
+        for i, fanout in enumerate(fanouts):
+            if fanout is None:
+                p_frontier = dgl.in_subgraph(g, p_curr)
+                n_frontier = dgl.in_subgraph(g, n_curr)
+            else:
+                p_frontier = dgl.sampling.sample_neighbors(g, p_curr, fanout)
+                n_frontier = dgl.sampling.sample_neighbors(g, n_curr, fanout)
+            p_block = dgl.to_block(p_frontier, p_curr)
+            n_block = dgl.to_block(n_frontier, n_curr)
+            p_curr = p_block.srcdata[dgl.NID]
+            n_curr = n_block.srcdata[dgl.NID]
+            p_blocks.insert(0, p_block.int())
+            n_blocks.insert(0, n_block.int())
```

Comment:
there is no removal of the predicted links?

Reply:
Yes, otherwise, it will be very slow.